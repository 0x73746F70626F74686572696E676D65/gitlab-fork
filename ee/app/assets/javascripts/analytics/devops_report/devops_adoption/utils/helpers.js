/**
 * A helper function which accepts the enabledNamespaces,
 *
 * @param {Object} params the enabledNamespaces data, and the check for open modals
 *
 * @return {Boolean} a boolean to determine if table data should be polled
 */
export const shouldPollTableData = ({ enabledNamespaces, openModal }) => {
  if (openModal) {
    return false;
  } else if (!enabledNamespaces.length) {
    return true;
  }

  const anyPendingEnabledNamespaces = enabledNamespaces.some(
    (node) => node.latestSnapshot === null,
  );

  return anyPendingEnabledNamespaces;
};
