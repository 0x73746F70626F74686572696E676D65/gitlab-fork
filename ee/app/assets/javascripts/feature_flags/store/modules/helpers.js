// eslint-disable-next-line import/prefer-default-export
export const parseFeatureFlagsParams = params => ({
  operations_feature_flags: {
    name: params.name,
    description: params.description,
    active: true,
    scopes_attributes: params.scopes.map(scope => ({
      environment_scope: scope.name,
      active: scope.active,
    })),
  },
});
