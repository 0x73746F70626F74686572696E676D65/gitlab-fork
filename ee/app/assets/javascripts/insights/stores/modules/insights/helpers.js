import { CHART_TYPES } from 'ee/insights/constants';
import { __ } from '~/locale';

const getGroupByValue = (queryParams) => {
  return queryParams.group_by || queryParams.params.group_by;
};

const getTypeValue = (queryParams) => {
  return queryParams.issuable_type || queryParams.params.issuable_type;
};

const getAxisTitle = (label) => {
  switch (label) {
    case 'day':
      return __('Days');
    case 'week':
      return __('Weeks');
    case 'month':
      return __('Months');
    case 'issue':
      return __('Issues');
    case 'merge_request':
      return __('Merge requests');
    default:
      return '';
  }
};

export const transformChartDataForGlCharts = (
  { type, query: queryParams },
  { labels, datasets },
) => {
  const formattedData = {
    xAxisTitle: getAxisTitle(getGroupByValue(queryParams)),
    yAxisTitle: getAxisTitle(getTypeValue(queryParams)),
    labels,
    datasets: [],
    seriesNames: [],
  };

  switch (type) {
    case CHART_TYPES.BAR:
      formattedData.datasets = [
        {
          name: 'all',
          data: labels.map((label, i) => [label, datasets[0].data[i]]),
        },
      ];
      break;
    case CHART_TYPES.STACKED_BAR:
      formattedData.datasets.push(
        ...datasets.map((dataset) => ({
          name: dataset.label,
          data: dataset.data,
        })),
      );
      break;
    case CHART_TYPES.LINE:
      formattedData.datasets.push(
        ...datasets.map((dataset) => ({
          name: dataset.label,
          data: labels.map((label, i) => [label, dataset.data[i]]),
        })),
      );

      break;
    default:
      formattedData.datasets = { all: labels.map((label, i) => [label, datasets[0].data[i]]) };
  }
  return formattedData;
};

export default {
  transformChartDataForGlCharts,
};
