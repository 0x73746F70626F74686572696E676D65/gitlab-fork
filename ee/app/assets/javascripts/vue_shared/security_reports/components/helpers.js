import { EMPTY_BODY_MESSAGE } from './constants';

/**
 * A helper function which validates the passed
 * in body string.
 *
 * It returns an empty string if the body has explicitly
 * been passed in as an empty string, a fallback
 * message if the body is null / undefined, else
 * it will return the original body string.
 *
 * @param {String} body the body message
 *
 * @return {String} the validated body message
 */
export const bodyWithFallBack = (body) => (body === '' ? '' : body || EMPTY_BODY_MESSAGE);

export const getCreatedIssueForVulnerability = (vulnerability) =>
  vulnerability.issue_links?.find((link) => link.link_type === 'created');

export const getDismissalTransitionForVulnerability = (vulnerability) => {
  const latestTransition = vulnerability.state_transitions?.at(-1);
  return latestTransition?.to_state.toLowerCase() === 'dismissed' ? latestTransition : null;
};
