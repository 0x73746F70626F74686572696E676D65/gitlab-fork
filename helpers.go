/*
Miscellaneous helpers: logging, errors, subprocesses
*/

package main

import (
	"fmt"
	"net/http"
	"os"
	"os/exec"
	"path"
	"syscall"
)

func httpError(w http.ResponseWriter, r *http.Request, error string, code int) {
	if r.ProtoAtLeast(1, 1) {
		// Force client to disconnect if we render request error
		w.Header().Set("Connection", "close")
	}

	http.Error(w, error, code)
}

// Git subprocess helpers
func gitCommand(gl_id string, name string, args ...string) *exec.Cmd {
	cmd := exec.Command(name, args...)
	// Start the command in its own process group (nice for signalling)
	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
	// Explicitly set the environment for the Git command
	cmd.Env = []string{
		fmt.Sprintf("HOME=%s", os.Getenv("HOME")),
		fmt.Sprintf("PATH=%s", os.Getenv("PATH")),
		fmt.Sprintf("LD_LIBRARY_PATH=%s", os.Getenv("LD_LIBRARY_PATH")),
		fmt.Sprintf("GL_ID=%s", gl_id),
	}
	// If we don't do something with cmd.Stderr, Git errors will be lost
	cmd.Stderr = os.Stderr
	return cmd
}

func cleanUpProcessGroup(cmd *exec.Cmd) {
	if cmd == nil {
		return
	}

	process := cmd.Process
	if process != nil && process.Pid > 0 {
		// Send SIGTERM to the process group of cmd
		syscall.Kill(-process.Pid, syscall.SIGTERM)
	}

	// reap our child process
	cmd.Wait()
}

// Borrowed from: net/http/server.go
// Return the canonical path for p, eliminating . and .. elements.
func cleanURIPath(p string) string {
	if p == "" {
		return "/"
	}
	if p[0] != '/' {
		p = "/" + p
	}
	np := path.Clean(p)
	// path.Clean removes trailing slash except for root;
	// put the trailing slash back if necessary.
	if p[len(p)-1] == '/' && np != "/" {
		np += "/"
	}
	return np
}
