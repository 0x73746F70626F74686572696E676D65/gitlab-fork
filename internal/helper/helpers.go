package helper

import (
	"errors"
	"log"
	"net/http"
	"net/url"
	"os"
)

func Fail500(w http.ResponseWriter, err error) {
	http.Error(w, "Internal server error", 500)
	LogError(err)
}

func LogError(err error) {
	log.Printf("error: %v", err)
}

func SetNoCacheHeaders(header http.Header) {
	header.Set("Cache-Control", "no-cache, no-store, max-age=0, must-revalidate")
	header.Set("Pragma", "no-cache")
	header.Set("Expires", "Fri, 01 Jan 1990 00:00:00 GMT")
}

func OpenFile(path string) (file *os.File, fi os.FileInfo, err error) {
	file, err = os.Open(path)
	if err != nil {
		return
	}

	defer func() {
		if err != nil {
			file.Close()
		}
	}()

	fi, err = file.Stat()
	if err != nil {
		return
	}

	// The os.Open can also open directories
	if fi.IsDir() {
		err = &os.PathError{
			Op:   "open",
			Path: path,
			Err:  errors.New("path is directory"),
		}
		return
	}

	return
}

func URLMustParse(s string) *url.URL {
	u, err := url.Parse(s)
	if err != nil {
		log.Fatalf("urlMustParse: %q %v", s, err)
	}
	return u
}

func HTTPError(w http.ResponseWriter, r *http.Request, error string, code int) {
	if r.ProtoAtLeast(1, 1) {
		// Force client to disconnect if we render request error
		w.Header().Set("Connection", "close")
	}

	http.Error(w, error, code)
}
