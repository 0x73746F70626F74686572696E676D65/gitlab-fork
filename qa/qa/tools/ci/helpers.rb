# frozen_string_literal: true

module QA
  module Tools
    module Ci
      module Helpers
        # Logger instance
        #
        # @return [Logger]
        def logger
          @logger ||= Gitlab::QA::TestLogger.logger(level: "INFO", source: "CI Tools")
        end
      end
    end
  end
end
