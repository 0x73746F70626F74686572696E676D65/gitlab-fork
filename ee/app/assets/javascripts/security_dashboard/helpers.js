import isPlainObject from 'lodash/isPlainObject';
import { s__ } from '~/locale';
import { convertObjectPropsToSnakeCase } from '~/lib/utils/common_utils';
import { ALL, BASE_FILTERS } from 'ee/security_dashboard/store/modules/filters/constants';
import { REPORT_TYPES, SEVERITY_LEVELS } from 'ee/security_dashboard/store/constants';
import { VULNERABILITY_STATES } from 'ee/vulnerabilities/constants';

const parseOptions = obj =>
  Object.entries(obj).map(([id, name]) => ({ id: id.toUpperCase(), name }));

export const mapProjects = projects =>
  projects.map(p => ({ id: p.id.split('/').pop(), name: p.name }));

export const initFirstClassVulnerabilityFilters = projects => {
  const filters = [
    {
      name: s__('SecurityReports|Status'),
      id: 'state',
      options: [
        { id: ALL, name: s__('VulnerabilityStatusTypes|All') },
        ...parseOptions(VULNERABILITY_STATES),
      ],
      selection: new Set([ALL]),
    },
    {
      name: s__('SecurityReports|Severity'),
      id: 'severity',
      options: [BASE_FILTERS.severity, ...parseOptions(SEVERITY_LEVELS)],
      selection: new Set([ALL]),
    },
    {
      name: s__('SecurityReports|Report type'),
      id: 'reportType',
      options: [BASE_FILTERS.report_type, ...parseOptions(REPORT_TYPES)],
      selection: new Set([ALL]),
    },
  ];

  if (Array.isArray(projects)) {
    filters.push({
      name: s__('SecurityReports|Project'),
      id: 'projectId',
      options: [BASE_FILTERS.project_id, ...mapProjects(projects)],
      selection: new Set([ALL]),
    });
  }

  return filters;
};

export const getFormattedSummary = (rawSummary = {}) => {
  if (!isPlainObject(rawSummary)) {
    return [];
  }
  // Convert keys to snake case so they can be matched against REPORT_TYPES keys for translation
  const snakeCasedSummary = convertObjectPropsToSnakeCase(rawSummary);
  // Convert object to an array of entries to make it easier to loop through
  const summaryEntries = Object.entries(snakeCasedSummary);
  // Filter out empty entries as we don't want to display those in the summary
  const withoutEmptyEntries = summaryEntries.filter(
    ([, scanSummary]) => scanSummary?.vulnerabilitiesCount !== undefined,
  );
  // Replace keys with translations found in REPORT_TYPES if available
  const formattedEntries = withoutEmptyEntries.map(([scanType, scanSummary]) => {
    const name = REPORT_TYPES[scanType];
    return name ? [name, scanSummary] : null;
  });
  // Filter out keys that could not be matched with any translation and are thus considered invalid
  return formattedEntries.filter(entry => entry !== null);
};

export default () => ({});
