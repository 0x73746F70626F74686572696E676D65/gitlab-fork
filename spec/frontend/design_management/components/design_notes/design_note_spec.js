import { ApolloMutation } from 'vue-apollo';
import { nextTick } from 'vue';
import { GlAvatar, GlAvatarLink, GlDisclosureDropdown, GlDisclosureDropdownItem } from '@gitlab/ui';
import { mountExtended } from 'helpers/vue_test_utils_helper';
import DesignNote from '~/design_management/components/design_notes/design_note.vue';
import DesignReplyForm from '~/design_management/components/design_notes/design_reply_form.vue';
import TimeAgoTooltip from '~/vue_shared/components/time_ago_tooltip.vue';

const scrollIntoViewMock = jest.fn();
const note = {
  id: 'gid://gitlab/DiffNote/123',
  author: {
    id: 'gid://gitlab/User/1',
    username: 'foo-bar',
    avatarUrl: 'https://gitlab.com/avatar',
    webUrl: 'https://gitlab.com/user',
  },
  body: 'test',
  userPermissions: {
    adminNote: false,
  },
  createdAt: '2019-07-26T15:02:20Z',
};
HTMLElement.prototype.scrollIntoView = scrollIntoViewMock;

const $route = {
  hash: '#note_123',
};

const mutate = jest.fn().mockResolvedValue({ data: { updateNote: {} } });

describe('Design note component', () => {
  let wrapper;

  const findUserAvatar = () => wrapper.findComponent(GlAvatar);
  const findUserAvatarLink = () => wrapper.findComponent(GlAvatarLink);
  const findUserLink = () => wrapper.findByTestId('user-link');
  const findReplyForm = () => wrapper.findComponent(DesignReplyForm);
  const findEditButton = () => wrapper.findByTestId('note-edit');
  const findNoteContent = () => wrapper.findByTestId('note-text');
  const findDropdown = () => wrapper.findComponent(GlDisclosureDropdown);
  const findDropdownItems = () => findDropdown().findAllComponents(GlDisclosureDropdownItem);
  const findEditDropdownItem = () => findDropdownItems().at(0);
  const findDeleteDropdownItem = () => findDropdownItems().at(1);

  function createComponent(props = {}, data = { isEditing: false }) {
    wrapper = mountExtended(DesignNote, {
      propsData: {
        note: {},
        noteableId: 'gid://gitlab/DesignManagement::Design/6',
        ...props,
      },
      data() {
        return {
          ...data,
        };
      },
      mocks: {
        $route,
        $apollo: {
          mutate,
        },
      },
      stubs: {
        ApolloMutation,
        GlDisclosureDropdown,
        GlDisclosureDropdownItem,
        TimelineEntryItem: true,
        TimeAgoTooltip: true,
        GlAvatarLink: true,
        GlAvatar: true,
        GlLink: true,
      },
    });
  }

  it('should match the snapshot', () => {
    createComponent({
      note,
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render avatar with correct props', () => {
    createComponent({
      note,
    });

    expect(findUserAvatar().props()).toMatchObject({
      src: note.author.avatarUrl,
      entityName: note.author.username,
    });

    expect(findUserAvatarLink().attributes()).toMatchObject({
      href: note.author.webUrl,
      'data-user-id': '1',
      'data-username': `${note.author.username}`,
    });
  });

  it('should render author details', () => {
    createComponent({
      note,
    });

    expect(findUserLink().exists()).toBe(true);
  });

  it('should render a time ago tooltip if note has createdAt property', () => {
    createComponent({
      note,
    });

    expect(wrapper.findComponent(TimeAgoTooltip).exists()).toBe(true);
  });

  it('should not render edit icon when user does not have a permission', () => {
    createComponent({
      note,
    });

    expect(findEditButton().exists()).toBe(false);
  });

  it('should not display a dropdown if user does not have a permission to delete note', () => {
    createComponent({
      note,
    });

    expect(findDropdown().exists()).toBe(false);
  });

  describe('when user has a permission to edit note', () => {
    it('should open an edit form on edit button click', async () => {
      createComponent({
        note: {
          ...note,
          userPermissions: {
            adminNote: true,
          },
        },
      });

      findEditButton().vm.$emit('click');

      await nextTick();
      expect(findReplyForm().exists()).toBe(true);
      expect(findNoteContent().exists()).toBe(false);
    });

    describe('when edit form is rendered', () => {
      beforeEach(() => {
        createComponent(
          {
            note: {
              ...note,
              userPermissions: {
                adminNote: true,
              },
            },
          },
          { isEditing: true },
        );
      });

      it('should open an edit form on edit button click', async () => {
        createComponent({
          note: {
            ...note,
            userPermissions: {
              adminNote: true,
            },
          },
        });

        findEditDropdownItem().find('button').trigger('click');

        await nextTick();
        expect(findReplyForm().exists()).toBe(true);
        expect(findNoteContent().exists()).toBe(false);
      });

      it('should not render note content and should render reply form', () => {
        expect(findNoteContent().exists()).toBe(false);
        expect(findReplyForm().exists()).toBe(true);
      });

      it('hides the form on cancel-form event', async () => {
        findReplyForm().vm.$emit('cancel-form');

        await nextTick();
        expect(findReplyForm().exists()).toBe(false);
        expect(findNoteContent().exists()).toBe(true);
      });

      it('hides a form after update mutation is completed', async () => {
        findReplyForm().vm.$emit('note-submit-complete', { data: { updateNote: { errors: [] } } });

        await nextTick();
        expect(findReplyForm().exists()).toBe(false);
        expect(findNoteContent().exists()).toBe(true);
      });
    });
  });

  describe('when user has admin permissions', () => {
    it('should display a dropdown', () => {
      createComponent({
        note: {
          ...note,
          userPermissions: {
            adminNote: true,
          },
        },
      });

      expect(findDropdown().exists()).toBe(true);
      expect(findEditDropdownItem().exists()).toBe(true);
      expect(findDeleteDropdownItem().exists()).toBe(true);
      expect(findDropdown().props('items')[0].extraAttrs.class).toBe('gl-sm-display-none!');
    });
  });

  it('should emit `delete-note` event with proper payload when delete note button is clicked', () => {
    const payload = {
      ...note,
      userPermissions: {
        adminNote: true,
      },
    };

    createComponent({
      note: {
        ...payload,
      },
    });

    findDeleteDropdownItem().find('button').trigger('click');

    expect(wrapper.emitted()).toEqual({ 'delete-note': [[{ ...payload }]] });
  });
});
