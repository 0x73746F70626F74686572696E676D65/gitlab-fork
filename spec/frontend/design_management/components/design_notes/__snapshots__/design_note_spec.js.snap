// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Design note component default should match the snapshot 1`] = `
<timelineentryitem-stub
  class="design-note note-form"
  id="note_123"
>
  <glavatarlink-stub
    class="gl-float-left gl-mr-3 link-inherit-color js-user-link"
    data-user-id="1"
    data-username="foo-bar"
    href="https://gitlab.com/user"
  >
    <glavatar-stub
      alt="avatar"
      entityid="0"
      entityname="foo-bar"
      shape="circle"
      size="32"
      src="https://gitlab.com/avatar"
    />
  </glavatarlink-stub>
   
  <div
    class="gl-display-flex gl-justify-content-space-between"
  >
    <div>
      <gllink-stub
        class="js-user-link link-inherit-color"
        data-testid="user-link"
        data-user-id="1"
        data-username="foo-bar"
        href="https://gitlab.com/user"
      >
        <span
          class="note-header-author-name gl-font-weight-bold"
        >
          
        </span>
         
        <!---->
         
        <span
          class="note-headline-light"
        >
          @foo-bar
        </span>
      </gllink-stub>
       
      <span
        class="note-headline-light note-headline-meta"
      >
        <span
          class="system-note-message"
        />
         
        <gllink-stub
          class="note-timestamp system-note-separator gl-display-block gl-mb-2 gl-font-sm link-inherit-color"
          href="#note_123"
        >
          <timeagotooltip-stub
            cssclass=""
            datetimeformat="DATE_WITH_TIME_FORMAT"
            time="2019-07-26T15:02:20Z"
            tooltipplacement="bottom"
          />
        </gllink-stub>
      </span>
    </div>
     
    <div
      class="gl-display-flex gl-align-items-flex-start gl-mt-n2 gl-mr-n2"
    >
       
      <div
        class="emoji-picker"
        data-testid="note-emoji-button"
      >
        <div
          boundary="viewport"
          class="dropdown b-dropdown gl-dropdown position-static btn-group"
          id="__BVID__15"
          lazy=""
          menu-class="dropdown-extended-height"
          no-flip=""
        >
          <!---->
          <button
            aria-expanded="false"
            aria-haspopup="menu"
            class="btn dropdown-toggle btn-default btn-md note-action-button note-emoji-button btn-icon btn-default-tertiary gl-button gl-dropdown-toggle btn-default-secondary"
            id="__BVID__15__BV_toggle_"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="award-control-icon-neutral gl-button-icon gl-icon gl-icon s16"
              data-testid="slight-smile-icon"
              role="img"
            >
              <use
                href="file-mock#slight-smile"
              />
            </svg>
             
            <svg
              aria-hidden="true"
              class="award-control-icon-positive gl-button-icon gl-icon gl-left-3! gl-icon s16"
              data-testid="smiley-icon"
              role="img"
            >
              <use
                href="file-mock#smiley"
              />
            </svg>
             
            <svg
              aria-hidden="true"
              class="award-control-icon-super-positive gl-button-icon gl-icon gl-left-3! gl-icon s16"
              data-testid="smile-icon"
              role="img"
            >
              <use
                href="file-mock#smile"
              />
            </svg>
          </button>
          <ul
            aria-labelledby="__BVID__15__BV_toggle_"
            class="dropdown-menu dropdown-extended-height"
            role="menu"
            tabindex="-1"
          >
            <!---->
          </ul>
        </div>
      </div>
       
      <!---->
       
      <!---->
    </div>
  </div>
   
  <div
    class="note-text md"
    data-qa-selector="note_content"
    data-testid="note-text"
  />
    
  <div
    class="awards js-awards-block gl-px-2 gl-mt-5"
  >
    <button
      class="btn gl-mr-3 gl-my-2 btn-default btn-md gl-button selected"
      data-emoji-name="briefcase"
      data-testid="award-button"
      title="You reacted with :briefcase:"
      type="button"
    >
      <!---->
       
      <!---->
       
      <span
        class="award-emoji-block"
        data-testid="award-html"
      >
        <gl-emoji
          data-name="briefcase"
        />
      </span>
       
      <span
        class="gl-button-text"
      >
         
        <span
          class="js-counter"
        >
          1
        </span>
      </span>
    </button>
    <button
      class="btn gl-mr-3 gl-my-2 btn-default btn-md gl-button selected"
      data-emoji-name="baseball"
      data-testid="award-button"
      title="You reacted with :baseball:"
      type="button"
    >
      <!---->
       
      <!---->
       
      <span
        class="award-emoji-block"
        data-testid="award-html"
      >
        <gl-emoji
          data-name="baseball"
        />
      </span>
       
      <span
        class="gl-button-text"
      >
         
        <span
          class="js-counter"
        >
          1
        </span>
      </span>
    </button>
     
    <div
      class="award-menu-holder gl-my-2"
    >
      <div
        class="emoji-picker"
        data-testid="emoji-picker"
        title="Add reaction"
      >
        <div
          boundary="scrollParent"
          class="dropdown b-dropdown gl-dropdown btn-group"
          id="__BVID__25"
          lazy=""
          menu-class="dropdown-extended-height"
          no-flip=""
        >
          <!---->
          <button
            aria-expanded="false"
            aria-haspopup="menu"
            class="btn dropdown-toggle btn-default btn-md add-reaction-button btn-icon gl-relative! gl-button gl-dropdown-toggle btn-default-secondary"
            id="__BVID__25__BV_toggle_"
            type="button"
          >
            <span
              class="gl-sr-only"
            >
              Add reaction
            </span>
             
            <span
              class="reaction-control-icon reaction-control-icon-neutral"
            >
              <svg
                aria-hidden="true"
                class="gl-icon s16"
                data-testid="slight-smile-icon"
                role="img"
              >
                <use
                  href="file-mock#slight-smile"
                />
              </svg>
            </span>
             
            <span
              class="reaction-control-icon reaction-control-icon-positive"
            >
              <svg
                aria-hidden="true"
                class="gl-icon s16"
                data-testid="smiley-icon"
                role="img"
              >
                <use
                  href="file-mock#smiley"
                />
              </svg>
            </span>
             
            <span
              class="reaction-control-icon reaction-control-icon-super-positive"
            >
              <svg
                aria-hidden="true"
                class="gl-icon s16"
                data-testid="smile-icon"
                role="img"
              >
                <use
                  href="file-mock#smile"
                />
              </svg>
            </span>
          </button>
          <ul
            aria-labelledby="__BVID__25__BV_toggle_"
            class="dropdown-menu dropdown-extended-height"
            role="menu"
            tabindex="-1"
          >
            <!---->
          </ul>
        </div>
      </div>
    </div>
  </div>
   
  <!---->
</timelineentryitem-stub>
`;
